xarray/tests/test_strategies.py

```diff
@@ -0,0 +1,325 @@
+        sub_seq_strat = subsequences_of(seq)
+        sub_seq = data.draw(sub_seq_strat)
+
+        # Per unique item, check that they don't occur in the subsequence
+        # more times that they appear in the source.
+        for item in set(sub_seq):
+            assert sub_seq.count(item) <= seq.count(item)
+
+    @given(st.data(), st.lists(st.integers()))
+    def test_subsequence_max_size_constraint(self, data, seq):
+        max_size_strat = st.integers(min_value=0, max_value=len(seq))
+        max_size = data.draw(max_size_strat)
+
+        sub_seq_strat = subsequences_of(seq, max_size=max_size)
+        sub_seq = data.draw(sub_seq_strat)
+
+        assert len(sub_seq) <= max_size
+
+    @given(st.data(), st.lists(st.integers()))
+    def test_subsequence_min_size_constraint(self, data, seq):
+        min_size_strat = st.integers(min_value=0, max_value=len(seq))
+        min_size = data.draw(min_size_strat)
+
+        sub_seq_strat = subsequences_of(seq, min_size=min_size)
+        sub_seq = data.draw(sub_seq_strat)
+
+        assert len(sub_seq) >= min_size
+
+    @given(st.data(), st.lists(st.integers()))
+    def test_subsequence_min_max_size_constraint(self, data, seq):
+        min_size_strat = st.integers(min_value=0, max_value=len(seq))
+        min_size = data.draw(min_size_strat)
+
+        max_size_strat = st.integers(min_value=min_size, max_value=len(seq))
+        max_size = data.draw(max_size_strat)
+
+        sub_seq_strat = subsequences_of(seq, min_size=min_size, max_size=max_size)
+        sub_seq = data.draw(sub_seq_strat)
+
+        assert min_size <= len(sub_seq) <= max_size
+
+    # this is a new test, important for keeping dimension names in order
+    @given(st.data(), st.lists(st.integers()))
+    def test_ordering_preserved(self, data, seq):
+        subsequence_of_dims = data.draw(subsequences_of(seq))
+        assert sorted(subsequence_of_dims) == subsequence_of_dims
+
+
+class TestCoordinateVariablesStrategy:
+    @given(coordinate_variables(dim_sizes={"x": 2, "y": 3}))
+    def test_alignable(self, coord_vars):
+
+        # TODO there must be a better way of checking align-ability than this
+        for v in coord_vars.values():
+            if "x" in v.dims:
+                assert v.sizes["x"] == 2
+            if "y" in v.dims:
+                assert v.sizes["y"] == 3
+            if not set(v.dims).issubset({"x", "y"}):
+                assert False, v
+
+    @given(st.data())
+    def test_valid_set_of_coords(self, data):
+        coord_vars = data.draw(coordinate_variables(dim_sizes={"x": 2, "y": 3}))
+
+        arr = data.draw(np_arrays(shape=(2, 3)))
+        da = DataArray(data=arr, coords=coord_vars, dims=["x", "y"])
+        assert isinstance(da, DataArray)
+
+    def test_generates_1d_dim_coords(self):
+        # TODO having a `hypothesis.find(strat, predicate)` function would be very useful here
+        # see https://github.com/HypothesisWorks/hypothesis/issues/3436#issuecomment-1212369645
```

Comment:
@Zac-HD specifically here I want to be able to check that my generated DataArrays produce 1D dimension coordinates at at least some of the time.

xref https://github.com/HypothesisWorks/hypothesis/issues/3436#issuecomment-1212369645

Reply:
Write an `@given()` test that fails if it gets a 1D dimension, disable shrinking, wrap it in `pytest.raises()` to invert the result.

Reply:
@Zac-HD having tried to write this test, I'm not sure I actually understand how to get the desired behaviour.

I need a test that passes if the strategy generates at least one example which meets `<condition>`, but all I can work out how to write is a test which ensures that all generated examples meet `<condition>`, which is too strong.

```python
@given(st.data())
@settings(phases=[Phase.generate])
def test_generates_1d_dim_coords(data):
    # TODO having a `hypothesis.find(strat, predicate)` function would be very useful here
    # see https://github.com/HypothesisWorks/hypothesis/issues/3436#issuecomment-1212369645

    coord_vars = data.draw(coordinate_variables(dim_sizes={"x": 2, "y": 3}))

    for name, var in coord_vars.items():
        with pytest.raises(AssertionError, match="Found a 1D dimension coordinate in generated examples"):
            if var.ndim == 1 and name == var.dims[0]:
                raise AssertionError("Found a 1D dimension coordinate in generated examples")
```

Could you please expand on how i should write this so that it only fails if no examples are generated that meet `<condition>`?

> wrap it in pytest.raises() to invert the result.

I feel like I need to invert the result of the entire parametrized test, not just the result of the assert. But I can only do that with `pytest.mark.xpass`, which isn't quite what I want either.

Reply:
```python
def test_sometimes_generates_1d_dim_coords():
    found_one = False

    @given(st.data())
    @settings(phases=[Phase.generate])
    def inner(data):
        coord_vars = data.draw(coordinate_variables(dim_sizes={"x": 2, "y": 3}))
        for name, var in coord_vars.items():
            if var.ndim == 1 and name == var.dims[0]:
                nonlocal found_one
                found_one = True 
                raise AssertionError  # early stopping - test is correct but slower without this

    with contextlib.suppress(AssertionError):
        inner()

    assert found_one
```

Reply:
Amazing! Works great.

This is pretty complex though - does make me wonder if the 
`hypothesis.find(<strat>, <predicate>)` mentioned would be worthwhile here.

xref https://github.com/HypothesisWorks/hypothesis/issues/3436#issuecomment-1212369645